File: rust/etsm/tests/virtual_call.rs
use etsm::*;

state_machine!(
    Foo,
    Data,
    State {
        // state, enter, exit, data
        {  A,     None,  None, Some(Data { run : Foo::run_a })},
        {  B,     None,  None, Some(Data { run : Foo::run_b })}
    }
);

struct Data {
    run: fn(&mut Foo),
}

struct Foo {
    state_machine: StateMachine<State>,
}

impl Foo {
    fn new() -> Foo {
        Foo {
            state_machine: StateMachine::new(),
        }
    }

    fn run_a(&mut self) {
        print!(" A ");
    }

    fn run_b(&mut self) {
        print!(" B ");
    }

    fn run(&mut self) {
        // call run on the current state
        if let Some(statedata) = get_statedata(&self.state_machine) {
            (statedata.run)(self);
        }
    }
}

#[test]
fn virtual_call() {
    let mut foo = Foo::new();

    transition!(&mut foo, state_machine, Some(State::A));
    foo.run();

    transition!(&mut foo, state_machine, Some(State::B));
    foo.run();

    // output: A B
}

File: cs/test/virtual_call.cs
using Microsoft.VisualStudio.TestTools.UnitTesting;
using Etsm;
using System;

namespace virtual_call
{
    // Define your own state and add Run Action.
    public class State : Etsm.State
    {
        public Action Run { get; private set; }

        public State(Action enter, Action exit, Action run = null)
            : base(enter, exit)
        {
            Run = run;
        }
    }

    public class Foo
    {
        private StateMachine<State> stateMachine;
        private State stateA;
        private State stateB;

        public Foo()
        {
            stateMachine = new StateMachine<State>();
            stateA = new State(null, null, RunA);
            stateB = new State(null, null, RunB);
        }

        private void RunA()
        {
            Console.Write(" A ");
        }

        private void RunB()
        {
            Console.Write(" B ");
        }

        // call run on the current state
        private void Run()
        {
            stateMachine.CurrentState?.Run?.Invoke();
        }

        public void Test()
        {
            stateMachine.Transition(stateA);
            Run();
            stateMachine.Transition(stateB);
            Run();
        }
    }

    [TestClass]
    public class virtual_call
    {
        [TestMethod]
        public void Test()
        {
            Foo foo = new Foo();
            foo.Test();
        }
    }
}
